### ISSUE TITLE
WantedBy=/Requisite= - fails to recognize the nonexistence of device units

### ISSUE THREAD

**Issue by thx1111** [systemd/systemd#5604]
- [ ] Bug report
  - [ ] Request for enhancement (RFE)

> systemd 233-1
> Arch

systemd does not know, a priori, that a `service` unit file may have the function of creating some `device` unit, or that, instead, another `service` unit file may not have the function of, itself, creating some `device` unit.

When a `service` unit  has a `WantedBy` dependency upon *both* some `device` unit and some other `service` unit, systemd will not be able to "interpret", ahead of time, based upon the  present-time existence or non-existence of a `device` unit, whether that  `service` unit should or should not be enqueued to start, possibly so as to create the `device` unit, without some mechanism which informs systemd about what action to take.

It may be that some `service` unit, rather than creating a `device` unit, has the function of simply configuring a `device` unit.  And then, that `device` unit might, or might not, actually already exist at any particular time.

In those circumstance where a `service` unit only configures a `device` unit, and does *not* create a `device` unit, where instead, for instance, a `device` unit is created directly by `systemd-udevd`, by plugging a removable device, say, then systemd must have some mechanism to determine that the `service` unit must *not* be enqueued to start, in a situation in which the `device` unit *does not yet exist*.

It would seem apparent, reading from `man systemd.unit`, that a `Requisite` dependency in the `service` unit, such that the `service` unit depends upon the `device` unit, would inform systemd that the `service` unit should not be enqueued to start when the `device` unit does not exist.  Reading from `man systemd.unit`, at `Requisite=`:
> if the units listed here are not started already, they will not be started
> and the transaction will fail immediately.

Instead, in systemd 233, a `service` unit `Requisite= sys-subsystem-net-devices-%p.device` fails to recognize when the listed `device` unit does not exist.  Of course, then, 
the `service` unit is started and then fails, with errors:
> command failed: No such device (-19)
> Cannot find device ...
> Main process exited, code=exited, status=1/FAILURE
> Failed to start ...
> Unit entered failed state.
> Failed with result 'exit-code'.

This is a bug and should be fixed.  `Requisite=` must properly recognize `device` units, when they exist, and when they *do not* exist, in determining the proper time at which to enqueue a `service` unit to start.

The `WantedBy` dependency does work correctly when a `device` unit appears, as when a device is plugged, or when another `service` unit is started.  But `WantedBy`, or the complementary `Wants`, must be "qualified" by, be made conditional upon, the `Requisite` dependency.

It might be argued that the `After` dependency could be used in place of the `Requisite` dependency.  Instead, the `After` dependency also fails as a condition for preventing the start of the `service` unit, *and* creates an additional problem, where consequent `service` units cannot start until after `DefaultTimeoutStartSec`, where the dependent `service` unit is "waiting", in futility, for the `device` unit to start.  In fact, this use of `After` can pointlessly block the entire boot process for `DefaultTimeoutStartSec`, and is a very wrong approach.

**Comment by arvidjaar** [comment]
> Requisite= must properly recognize device units, when they exist, and when they do not exist, in determining the proper time at which to enqueue a service unit to start.

Was not it already discussed in one of your previous reports? All that `Requisite` does is to submit additional job to test whether other unit is active. The only ordering between jobs (as implemented currently) is that imposed by `After` and `Before` of units. So unless `Requisite` is complemented with `After` both `start` and `verify` jobs run concurrently.

So yes, both `Requires` and `Requisite` as implemented currently are mostly useless without `After` and I have not yet seen example where they would make sense without `After`.

**Comment by thx1111** [comment]
> Was not it already discussed in one of your previous reports?

No, not in this context.  The symptom, failing to recognize the non-existence of a *device*, and then pointlessly running a service unit file, is still a problem that has *not* been addressed.

In particular, with respect to your description of 'Requisite', a non-existent device is *not active*.  By that simple measure, the `Requisite` test has demonstrated a bug.

This is *not* an *ordering* issue, and so has nothing to do with `Before` or `After`.  Instead, we might say that this is a "condition" issue.

Note - a *device* unit is not like other units, in the sense that device units "become active" and "become inactive", *on their own*, without absolute control by systemd.  Instead, systemd must *respond* to the spontaneous "appearance" and "disappearance" of device units.

Here, the problem occurs when the `WantedBy` dependency between two *service* units causes the first service unit to "start" the second service unit - which itself has the function of configuring a device - but this is occurring *when that device does not exist*.  In this case, *only* the action of `WantedBy` is being recognized, and systemd utterly fails to recognize and include the second service unit's `Requisite` constraint, meaning the existence of the device.  This is a bug.

There are *three* units interacting in this scenario: the first service unit, the second service unit, and the device unit.  The first service unit `Wants` the second service unit, but the second service unit is `Requisite` upon the device unit.  The second service unit should not be enqueued for start until the `Requisite` dependency is satisfied, meaning, until, and when, the device unit becomes active.

Also, note here that this has nothing to do with the idea of "After a device", which would be a meaningless expression.  A "device unit" is NOT a kind of unit with an "Executable program" that can be "run" or "managed" by systemd!  A "device unit" is *not* like a "service unit"!  systemd does not have the ability to make a device "active", or to make a device "inactive"!  Yet still, systemd *must* be able to *recognize* that a device is active or inactive.

**Comment by poettering** [comment]
I am not sure I follow. but it appears to me that your device unit should pull in via a udev rule using SYSTEMD_WANTS= the service that configures it, and then your main service should simply order itself after that... that way, if the device is around the configuration service for it is around too, but you sitll get the correct syncing if you need it...

**Comment by thx1111** [comment]
Yes, that function works just as usual.  The problem is that when *some other service unit*, rather than the device itself, "pulls in" the service unit, and *that device does NOT exist*, then systemd will *still* enqueue the service unit for start.

You keep talking about "if the device is around".  That is not the circumstance being addressed here.  The circumstance is "if the device is NOT around".  In that case, systemd fails.  It is a bug.

Closing the issue does not make the bug go away.  `Requisite=` on a device is still broken.

**Comment by poettering** [comment]
> The problem is that when some other service unit, rather than the device itself, "pulls in" the service unit, and that device does NOT exist, then systemd will still enqueue the service unit for start.

Sure, if you pull in the service and that service wants to wait for the device, then things will block, and that's expected. If you don't want that behaviour then don't add a dependency on the service and all should be good. You shouldn't place deps you don't want.

**Comment by thx1111** [comment]
> and that service wants to wait for the device, then things will block

No, the service does NOT want to "wait" for the device, specifically because we do NOT want to block - and lock-up the entire computer system.

> don't add a dependency on the service and all should be good

That is an amazingly rude suggestion.

The `Requisite=` dependency, when that dependency is upon a device unit, fails to function.  It is broken.  There is a bug.  You've heard the joke about the guy who goes to see the doctor about motion pain?  The guy says "It hurts when I move like this."  And the doctor replies "Then don't move like that."  When you say it, though, it's not funny.

A service unit with a `BindsTo=unit.device` dependency does not "wait" for the device unit, but it does become enqueued for start when that device unit "appears".  That part works fine.  Now, suppose that you want all the slaves devices to be configured only *if* the master device is being enabled, but where some of the slave devices do not exist?  The slave devices come and go, and the configuration has to respond accordingly.  What to do?  Easy - you condition on the existence of the slave device - except systemd `Requisite` is broken.

Supposedly, this broken "conditioning" behavior was going to be "fixed" using `BindsTo` and `After` in combination.  But that "fix" utterly failed to do anything useful, and only generates meaningless and misleading warnings in the journal.  It was also a bad idea, trying to mix "conditioning" and "ordering".

I really do not understand why you are so resistant to simply making `Requisite=` work correctly with a device type unit.  It's all of, like, a single if-else:

 if ( service unit is `Requisite` device unit && device unit is NOT active )
     bail-out, do NOT enque the service unit for start, and be done with it ;
 else
    { continue doing what you were already going to do ; }

Capiche?

**Comment by poettering** [comment]
> don't add a dependency on the service and all should be good
> 
> That is an amazingly rude suggestion.

Well, you think that systemd should work a specific way, a way it really doesn't work and shouldn't work by design. You may consider that a bug, but we really do not, we consider it an incorrect usage, and are just suggesting you shouldn't add those deps the way you are doing, because that's not how you are supposed to use them according to our design.

> Capiche?

And that's not rude?

**Comment by poettering** [comment]
> if ( service unit is Requisite device unit && device unit is NOT active )
> bail-out, do NOT enque the service unit for start, and be done with it ;
> else
> { continue doing what you were already going to do ; }

That's not how this is supposed to work. all of systemd's dependencies that propagate failures will do so honouring the ordering deps in place, i.e. are tested as result of the job queue being processed. The job queue's ordering is strictly enforced, and we check failure only at the moment a job is getting to the point to being processed.

**Comment by thx1111** [comment]
> we consider it an incorrect usage, and are just suggesting you shouldn't add those
> deps the way you are doing, 

The documentation says otherwise.

I've been running this around in my head for about a year now, and the only way to have this work - without `Requisite` - appears to be to just put up with systemd starting the service units, and letting the service units run on the nonexistent devices.

>> Capiche?

> And that's not rude?

Maybe - between the two of us, being the way we are...

> all of systemd's dependencies that propagate failures will do so honouring
> the ordering deps in place

There is nothing wrong with the ordering.  systemd does that correctly.  `Requisite` is not an ordering dependency, though.

> That's not how this is supposed to work.

Have you actually *read* the documentation for `Requisite`?

 `if the units listed here are not started already, they will not be started and the transaction
 will fail immediately.`

You seem to have some unusual interpretation of "unit ... not started already".  You seem to be implying a meaning which is distinct from "device unit is NOT active".  Can you clarify this difference, between a "device that is not active" and a "device that is not already started"?

I am assuming here that a "non-existent" device is a device that can not be "active",  and that can not be "started already".

And then, what would be the meaning of an "already started non-existent device"?  That does not seem to make any sense.

Or, am I misunderstanding the meaning of "units ... will not be started"?  I take that to mean that a service unit will not be enqueued for start, "not be started".  Does that have some other meaning?

**Comment by arvidjaar** [comment]
> we consider it an incorrect usage

So using program according to documention is now considered incorrect usage? Documentation says that if A has `Requisite=B`, A won't be started if B is not already started. That obviously does not happen here - A is attempted to be started even though B is not active. Please explain what user does wrong?

**Comment by vad1m** [comment]
@poettering ?

**Comment by poettering** [comment]
> So using program according to documention is now considered incorrect usage? Documentation says that if A has Requisite=B, A won't be started if B is not already started. That obviously does not happen here - A is attempted to be started even though B is not active. Please explain what user does wrong?

I am not sure where the confusion is, but Requisite=B in a service A will *not* cause B to be started if A is started. Instead, starting A will enqueue a JOB_START job for A, and a JOB_VERIFY_ACTIVE job for B, which will dispatched according to the order in effect between A and B. Normally, if you have After=B in A, then this means the JOB_VERIFY_ACTIVE job will be run first: it will fail if the service is not running, and this will then be propagated from B to A and A will fail too.

Now, there's a concept of job merging in systemd: if three clients want X to start at the same time, then their three requests are merged and only a single job for X is enqueued. This merging also works in limited ways between different job types: if one client enqueues a JOB_START job and the other a JOB_VERIFY_ACTIVE job for the same unit these requests get merged, and the "stronger" one is the one that is enqueued, which is JOB_START. Other job types can not be merged however: JOB_START and JOB_STOP directly contradict each other, and hence are never merged, and enqueing will fail.

Now, @thx1111 enqueued JOB_VERIFY_ACTIVE through the Requisite= dep, and JOB_START through his other dep from another service, then the result will be that JOB_START is queued, and the JOB_VERIFY_ACTIVE job is merged into it.

Why is this done that way? becasue we minimize work, and systemd tries to "think positive", i.e. you asked systemd to do two things and even defined an order between them, and one would fail without the other, hence of course we'll try to execute things in the way we can best to reach the desired result.

Hence again: Requisite= alone will never pull in another unit. However, if you pull the other unit in via some other means (which @thx1111 did) then we'll make the best of it.

My suggestion is still: if you don#t want to pull in a specific unit, then don't pull it in.

**Comment by thx1111** [comment]
> I am not sure where the confusion is

Clearly!  Please take some time to rethink this issue, and to think it through...

> ...but Requisite=B in a service A will not cause B to be started if A is started.

That statement is entirely off-point and "out of left field".  What inspired you to go off talking about starting Unit B, when the intended and seemingly obvious topic is the starting of Unit A???

And then, you continue with "... starting A ...", which is to "presume the conclusion", that Unit A is being started.  That Unit A will be enqueued for start is NOT a valid assumption of the issue under discussion, but is instead the issue being addressed.

>  which will dispatched according to the order in effect between A and B.

That would be an obvious "programming design error", since then, "ordering" and "dependency" would NOT be orthogonal, when the statement has been made, by you, many many times, that "ordering" and "dependency" ARE orthogonal.

Again, device units cannot be directly "enqueued for start" by systemd.  Device units are either active or they are not active.

Ordering will take care of itself.  Yes, *if* systemd were able to directly enqueue some Unit B, and Unit B were to be already enqueued for start, supposing that Unit B were not already active, and where a "B starts after A" dependency has been specified , *then* we *might* be inclined to allow that "enqueued for start" is equivalent to "already active", but that is a stretch.  Normally, the reasonable connotation would be that the phrase in the documentation, "are not started already" means the same as "already active", and NOT also "enqueued for start".

Otherewise, ORDERING WILL NOT MATTER.  If Unit B is not already active, then Unit A should NOT be enqueued for start, no matter what ordering has been specified.  Supposing that Unit B might or might not, or could or could not, be started by systemd is a "red herring" in this conversation.  It is completely off point.

> Normally, if you have After=B in A, then this means the JOB_VERIFY_ACTIVE job will be
> run first: it will fail if the service is not running, and this will then be propagated from
> B to A and A will fail too.

That statement does NOT describe the actual systemd behavior.  Adding an explicit After=B in Unit A will NOT prohibit Unit A from being enqueued for start.  If an "After=B" was suppose to have some effect, then that is another bug.  Since ordering should NOT matter, then it is better that it does not matter, with respect to the way systemd actually works.

> if one client enqueues a JOB_START job and the other a JOB_VERIFY_ACTIVE job for
> the same unit these requests get merged, and the "stronger" one is the one that
> is enqueued, which is JOB_START.

Again, that statement is off-point.  You have raised the issue of enqueuing for start Unit B, to which "a JOB_VERIFY_ACTIVE job for B" would be applied, when the issue under discussion is with respect to enqueuing Unit A for start.  Please try to focus on the issue at hand.  And again, device units cannot be enqueued for start.

> enqueued JOB_VERIFY_ACTIVE through the Requisite= dep, and JOB_START through
> his other dep from another service, then the result will be that JOB_START is queued,
> and the JOB_VERIFY_ACTIVE job is merged into it.

Again, you are apparently talking about Unit B instead of Unit A.

> Hence again: Requisite= alone will never pull in another unit.

Nowhere in this thread was there any suggestion that Requisite= would "pull in another unit", where "pull in " presumably means to "enque for start", and implicitly, Unit B.

Perhaps, by "pull in", you could mean "JOB_VERIFY_ACTIVE", but then that interpretation would make no sense, that "Requisite= alone will never set JOB_VERIFY_ACTIVE", since you  said earlier that that was the action of Requisite=, to set "a JOB_VERIFY_ACTIVE job for B".  So, I do not understand what you are trying to say there.

You seem to go out of your way to create a "straw man argument" with respect to Unit B, all the while avoiding a discussion with respect to Unit A.  It is difficult for me to fathom why you are doing this.  Has the systemd logic become such a tangled web that you are no longer able to grasp it all?  You might reconsider my proposal, 2016 Dec 23, to refactor the dependency functions into separate forward and reverse dependency functions, to simplify the logic and avoid confusion with respect to the operation of each dependency.  Trying to manage these "combined" dependencies can become unnecessarily convoluted.

And still, you have failed to respond to the issue raised, that the actual operation of the Requisite= dependency does not conform to the documentation.

**Comment by poettering** [comment]
> Again, device units cannot be directly "enqueued for start" by systemd. Device units are either active or they are not active.

Of course they can. You can do "systemctl start dev-foo.device" at any time, and it will wait for /dev/foo to appear. systemd might not be able to make the device appear but it can sure wait for it, expecting that something else makes it appear, and that's what happens.

**Comment by poettering** [comment]
I am sorry, but I really have trouble following what you try to say. Maybe I am not smart enough to follow, but I really don't understand, and I am not sure this discussion will lead to anything. We certainly have a communications problem here, and I am unable to clear it up. Sorry. If somebody else groks what you are trying to say, and can put it in different words maybe I'll understand, but otherwise I fear this will lead to nothing. Sorry.

**Comment by arvidjaar** [comment]
@poettering 
> If somebody else groks what you are trying to say, and can put it in different words maybe I'll understand

I tried to do it multiple times. Let's try once more.

1. Documentation says for `Requisite`: "if the units listed here are not started already, they will not be started and the transaction will fail immediately".

2. In reality if user puts `Requisite` without corresponding `After` it causes race condition between two jobs - job that starts first unit and job that checks second unit. So sysemd still attempts to start unit even though dependent unit is not active.

3. This race condition without `After` is the main point of this report. Documentation says `Requisite` is enough to prevent unit from starting if dependent unit is not active, while in reality both `Requisite` and `After` must be present for it to be effective.

You may repeat ad nauseam that this is how systemd works, but this is **NOT** how it is **DOCUMENTED** to work. And users use software based on its documentation, not based on its source code.

**Comment by thx1111** [comment]
And, in my experience, in use, I have NOT seen that `Requisite=` actually has any effect whatsoever, whether or not a corresponding `After=` is included, certainly not with respect to a *device* unit, which, in all cases, systemd can not start and can not stop.  A "device unit" is NOT, and often does not have, a "device unit *configuration file*".

> You can do "systemctl start dev-foo.device" at any time, and it will wait for /dev/foo
> to appear. systemd might not be able to make the device appear but it can sure wait
> for it, expecting that something else makes it appear, and that's what happens.

What would be the practical purpose or value in having systemd "wait" for a device?

In my experience, having systemd "wait" for a device simply causes systemd to "lock-up", waiting for a device which will never "appear", which has about the same practical effect as having the OS crash, which is to say, that having systemd "wait" for a device is a very bad thing, and a good way to cause a "denial of service" with systemd.

Generally speaking, network devices do not have any "/dev/foo".  I don't know if you meant that example literally, as opposed to, say, "sys-subsystem-net-devices-%p.device".  Are you confusing, as a class, block devices with network devices?

**Comment by arvidjaar** [comment]
> What would be the practical purpose or value in having systemd "wait" for a device?

You cannot mount filesystem until its block device is present. You cannot configure interface until this interface is present.

**Comment by thx1111** [comment]
> You cannot mount filesystem until its block device is present.

Thanks.  That example illuminates two subtly different "start" sequences, distinguishing between fixed and removable devices.

On boot, with a fixed block device, there is nothing that can be done by systemd until the root block device and root filesystem are present.  Still, it is not immediately obvious how the block device is going to "magically" appear, though presumably by "waiting" for the kernel, by itself, to load the block device driver and then create the block device.

I would not think that an explicit "systemctl start dev-foo.device" would be necessary in that case, where instead, a simple `After=dev-sda.device` would suffice.

*But, it is not so simple with a removable/pluggable device.*

> You cannot configure interface until this interface is present.

Here, there are certain "virtual" network interfaces, such as bridge and bond interfaces.  These interfaces may be "started" - or *must* be "started" - by systemd.  In that case, systemd must first "start" the virtual network device, and only then, systemd could "wait" for that virtual network device to appear.  But then, there would seem to be no point in "waiting" for the network interface, once the network interface had been "started", except that "starting" the device might subsequently fail.  Even then, an `OnFailure=` dependency could be defined, and there would still be no point in "waiting" for the virtual network interface device with an `After=` ordering.

The start sequence for a pluggable device may be even more complex, though.  If the virtual network interface is itself composed of removable/pluggable "physical" network devices, then "waiting" for the physical network device with `After=`, before "starting" the corresponding virtual network device,  is inappropriate, where that physical network device might never be "plugged", or might be "plugged" at some time far in the future.  Instead, a `BindsTo=` dependency upon the physical network interface is used, as the condition for "starting" the corresponding virtual network device.

Of course `BindsTo` and `After` are quite different things.

And this is where the trouble with `Requisite=` occurs, where `BindsTo` alone is not sufficient to prevent systemd from enqueuing the unit file commands which "start" the virtual network interface device.  Unfortunately, `Requisite= physical.device` is also completely ineffective here.

And also, the recent change to `BindsTo` that modifies its behavior with `After=` is completely brain-dead, as can be seen by noting that the change does NOT solve the problem with "starting" the virtual network device when the physical network device does not exist, and by watching the consequent meaningless and contradictory journal messages.  Perhaps that is a separate issue...

Looking back, in Issue 4756, I see that Lennart uses the expression "you enqueue a check that a unit is started up properly, but then don't wait for the result of the check".  This idea of "enqueuing a check" is something that is counterintuitive and seems like a mistaken approach.

To "enqueue a check" implies that a "check" is going to be performed at some time in the future - and that is "crazy making", not an action that conforms to the reasonable meaning of the term "check".  It makes no sense to "check" some conditional state in the future.  That implies a time inversion.  I suspect that this idea of a "check in the future" is the basis of this broken `Requisite` dependency.

I'm not sure how better to express this.  You cannot have a "check in the future"!  That is not even a foundation for a meaningful conversation.  In this case, it is NOT appropriate to "shoot first and ask questions later".

To state the seemingly obvious, when an action is conditional upon some state, the condition of that state must be "checked" ***before*** the action is initiated.  A conditional action that is initiated before the condition itself is "checked" is the definition of a broken conditional action.

Yes, there is the difficulty of defining the "proper time" in a sequence of events, where here, there is the "time" in a queue of "start" and "stop" events which might suggest a "time in the future", and also there is the "time" of conditional states, which necessarily means a "time right now".  It appears that systemd is confusing these two kinds of "time".  Even then, this is no excuse for systemd initiating an action *before* a "check" has been completed - ever!

`Requisite` must reference a "time right now".

The idea of a "will have been activated in the future" dependency, referring to some group of enqueued unit "start" and "stop" processes in the future, as opposed to an "is now active" dependency, referring to a current-time single unit activation state, is something completely different.

What is the use case for this "other unit which will have been activated in the future" scenario, which Lennart seems to be addressing and assuming?

**Comment by arvidjaar** [comment]
> distinguishing between fixed and removable devices ... there are certain "virtual" network interfaces, such as bridge and bond interfaces

There is no difference. Either device is present or not; how it became present is irrelevant for the purpose of ordering unit jobs which is what we are discussing here.

> BindsTo alone is not sufficient to prevent systemd from enqueuing the unit file commands which "start" the virtual network interface device

That's bug in `BindsTo` design which conflates two entirely different things - propagating of start request and propagating of "surprise removal" event.

> "you enqueue a check that a unit is started up properly, but then don't wait for the result of the check". This idea of "enqueuing a check" is something that is counterintuitive and seems like a mistaken approach.

Ordering between units is implemented as ordering between jobs queued for these units. You have been told it more than once. What is so hard to understand here? To implement `A After B` you **must** have _some_ job queued for both A and B where job for A waits until job for B completes. And at the time you submit jobs for both A and B none of them is active yet so of course both jobs are in the future. Any event loop is about future events.

You are welcome to suggest different implementation that would also tolerate non-deterministic order in which units become active (where "units" include "devices" here).

**Comment by thx1111** [comment]
> Either device is present or not; how it became present is irrelevant for the purpose
> of ordering unit jobs which is what we are discussing here.

Agreed.  But it is not so much "ordering" as - shall we say - "dependency" that we are discussing.

> That's bug in BindsTo design which conflates two entirely different things -
> propagating of start request and propagating of "surprise removal" event.

Agreed.  The `BindsTo` response to the "surprise removal" event seems to be correct, where the depending unit is simply stopped.  For the behavior of `Requisite`with respect to propagating the start request, my original thought, before exploring the use of `Requisite=`, was that the `BindsTo` behavior was wrong.  My understanding and expectation was that `BindsTo` provided an "edge sensitive" event response, only responding to the "appearance" of a device unit, and that it should not start a unit when the depended-upon `BindsTo` device unit did not exist.  If `BindsTo` *did* work that way, the `Requisite` bug would not be an issue.  The new `BindsTo` connection to `After=` does nothing to correct the behavior of the `BindsTo` when propagating a start request.  The actual behavior of `BindsTo` seems to contradict the new documentation, which says "the unit bound to strictly has to be in active state for this unit to also be in active state".  In actuality, when "the unit bound to" does not exist, "this unit" is enqueued for start, nonetheless.

> Ordering between units is implemented as ordering between jobs queued for these
> units. You have been told it more than once. What is so hard to understand here?

It is important that I answer this question, and not simply leave it as rhetorical.  

> ... Any event loop is about future events.

You are correct that I have had to shift my point of view about the "proper time", to understand in terms of the event queue, rather than some "proper time" prior to the events in the queue.

Still, I will say that the `Requisite` conditional start mechanism is not working properly.  Let us suppose that a check for the activation state of a depended-upon unit is entered into the job queue, to be completed at some time in the future.  It is still an error to activate the depending unit before this check has been completed.

If both job A and job B are entered into the queue, job A must not be "enqueued for start", but might be in some state "enqueued for future condition check".  Job B must not be in some state "enqueued to wait for a long time for some unit B to appear", but instead should be in a state "enqueued to check immediately if unit B is currently active".

To address non-deterministic ordering, the systemd ordering mechanism still works just fine.  *If* it is desired that the job B activation state check should be executed after waiting for some job B to "have been started", then an `After= jobB` dependency can be defined.  And if it is desired that the job B activation state check be executed immediately, then that will happen when there is no `After= jobB` dependency.  I don't see a problem there.

The error would seem to be in the idea that job A should be entered into the queue at all before the job B "enqueued to check immediately if unit B is currently active" job has been completed.

Instead, a `Requisite` dependency in job A should block job A from being entered into the queue as a "start" job, and then enqueue a job that includes two variables, both "job A" and "job B", as "check immediately if unit B is currently active, and if unit B is active, then start unit A".  Here, the "if unit B then unit A" process is made into a "job" that can then be entered into the job queue.

Without looking at the code, I suspect that systemd jobs are not set up to accept two unit variables.  Would that be a problem?  An alternative would be to set up some kind of call-back mechanism.  But there still has to be some mechanism to connect job B with job A, independent of the systemd ordering mechanism.

Again, without looking at the code, I suspect that there already exists some mechanism from the `BindsTo` dependency that would allow job A to be entered into the queue in some "suspended start" state.  Then again, it seems that the `BindsTo` start mechanism is itself currently broken.  These two bugs, in `BindsTo` and `Requisite` may be related, or may be, effectively, the same bug.

I'm interested in your thoughts about these mechanisms, assuming that you are more familiar with the code base.

**Comment by arvidjaar** [comment]
> The error would seem to be in the idea that job A should be entered into the queue at all before the job B "enqueued to check immediately if unit B is currently active" job has been completed.

You confuse "queuing" with "selecting for execution".

> Instead ... "check immediately if unit B is currently active, and if unit B is active, then start unit A".

Which is exactly what happens now as long as A has both `Requisite=B` and `After=B`.

**Comment by thx1111** [comment]
> You confuse "queuing" with "selecting for execution".

Yes.  Please clarify how these are different things.

> Which is exactly what happens now as long as A has both Requisite=B and After=B.

Hmm - after again testing different combinations of `Requisite=B.device` and `After=B.device`, and looking more carefully at the journal messages, I see that the problem is a little more subtle.  In the case that you propose, with both `Requisite=B.device` and `After=B.device`, "technically" you are correct that job A is not actually "started", but there is still an error, in the sense of an intolerable timeout, waiting for a nonexistent B.device, which hangs systemd during boot.  In the case of neither `Requisite=B.device` nor `After=B.device`, or of only `Requisite=B.device`, the error is instead an actual "Failed with result 'exit-code'", when job A is actually "started".

So the idea that job A has been "started" or "enqueued" or "selected" or whatever is actually two different things.  And so really, I need to be describing two different kinds of problems.  You can help with the terminology here, to describe the state of job A while job A waits for a nonexistent B.device, and to describe the state of job A when job A is run with a nonexistent B.device.

I think that this might clarify the confusion, since I have been claiming that job A was being "started", which is not actually precise, and Lennart has been claiming that job A was NOT being "started", which is sort of correct but also not actually precise.  The problem that I am seeing with `Requisite`, then, is that it "waits" on B.device, which is unacceptable.

I still assert that explicitly requiring an `After=B.device` with `Requisite=B.device` is an error, and that "starting" job A from `BindsTo=B.device` with a nonexistent B.device is also an error.  But then those are arguably separate issues.

The facts remain:
* B.device does not exist on boot.
* `BindsTo=B.device` in unit file A - in some sense - "starts" job A even when B.device does not exist.
* `Requisite=B.device` in unit file A does NOT prevent job A from starting.
* `Requisite=B.device` in combination with `After=B.device` in unit file A does NOT prevent job A from waiting for a nonexistent B.device.
* `After=B.device` in unit file A will cause job A to hang systemd for `DefaultTimeoutStartSec` during boot.

The `DefaultTimeoutStartSec` hang can be avoided by simply removing `After= B`, and the other errors are not fatal, just annoying.

The journal messages are different when `Requisite=B.device`, and when `Requisite=B.device` in combination with `After=B.device`, are added to unit file A.

With `Requisite=B.device` in combination with `After=B.device`, there is no explicit "systemd[1]: Starting job A", and the error is "Timed out waiting for device ..." and "Dependency failed ...".  Without both `Requisite=B.device` and `After=B.device`, instead, the error is "Failed to start ..." and "Failed with result 'exit-code'".  For example:

 systemd[1]: sys-subsystem-net-devices-wlp0s26u1u2.device: Job sys-subsystem-net-devices-wlp0s26u1u2.device/start timed out.
 systemd[1]: Timed out waiting for device sys-subsystem-net-devices-wlp0s26u1u2.device.
 systemd[1]: Dependency failed for Enslaved wlp0s26u1u2@bond0.
 systemd[1]: wlp0s26u1u2@bond0.service: Job wlp0s26u1u2@bond0.service/start failed with result 'dependency'.
 systemd[1]: sys-subsystem-net-devices-wlp0s26u1u2.device: Job sys-subsystem-net-devices-wlp0s26u1u2.device/start failed with result 'dependency'.

versus:

 systemd[1]: wlp0s26u1u2@bond0.service: Main process exited, code=exited, status=1/FAILURE
 systemd[1]: Failed to start Enslaved wlp0s26u1u2@bond0.
 systemd[1]: wlp0s26u1u2@bond0.service: Unit entered failed state.
 systemd[1]: wlp0s26u1u2@bond0.service: Failed with result 'exit-code'.

Additionally, without `After=B.device`, there are, for instance, repeated four times, the misleading error messages:

 systemd[1]: Unnecessary job for sys-subsystem-net-devices-wlp0s26u1u2.device was removed.

Whether this message refers to "job A" or to the job "checking for B.device" is unclear, but either way, job A will subsequently be "started", despite some "unnecessary job" having been "removed".

Additionally, without `Requisite=B.device`, there are, for instance, the following sequence of messages:

 systemd[1]: Started bridge0 Interface Master.
 systemd[1]: master@bridge0.service: Unit is bound to inactive unit sys-subsystem-net-devices-bridge0.device. Stopping, too.
 systemd[1]: Requested transaction contradicts existing jobs: Transaction is destructive.
 systemd[1]: master@bridge0.service: Failed to enqueue stop job, ignoring: Transaction is destructive.

I don't know what the interaction is here with `Requisite=B.device`, but these messages appeared after the modifications to `BindsTo`, from the previous version of systemd.  These messages may or may not be of any interest with respect to understanding the behavior of `Requisite`.

**Comment by arvidjaar** [comment]
> with both Requisite=B.device and After=B.device, "technically" you are correct that job A is not actually "started", but there is still an error, in the sense of an intolerable timeout, waiting for a nonexistent B.device

`Requisite` queues `JOB_VERIFY_ACTIVE` which has no timeout (it is completed as soon as it is selected for execution). In your case something else already waits for device, which causes `JOB_VERIFY_ACTIVE` to be subsumed by this waiting job.

This merging of two jobs is logical for normal type of units which have start method (if we are in process of starting unit, we should delay answer whether unit is active until starting has either succeeded of failed). It is indeed questionable in case of device unit where `start` is misused for `wait for device to appear`. Unfortunately I do not see any easy way to distinguish between devices which need some explicit action to appear (consider crypto device or your initial example of layered interfaces) and devices are just waited for.

So far nothing you describe is a bug (in the sense - behaves not as designed). It is poorly documented and we may question design, but behavior is consistent.

So I would suggest - you step back, try to describe what you are trying to achieve in general terms (without referring to any specific systemd directives) and do it on mailing list. Then we hopefully either find how to implement desired behavior, or at least identify what functionality is missing in systemd.

**Comment by thx1111** [comment]
> In your case something else already waits for device

Yes.  With `After=B.device` in unit file A, then unit A waits for B.device.  Would you agree?

Whether the terminology "started" is or is not appropriate for the state of unit A in this case is unclear.  Unit A is having a direct impact on the operation of systemd, even though the `ExecStart` commands are not yet being run.  And, when unit A just happens to include `Before= network.target`, this will block the entire boot process.

> Unfortunately I do not see any easy way to distinguish between devices which need
> some explicit action to appear (consider crypto device or your initial example of
> layered interfaces) and devices are just waited for.

Simply do not use the `After=` mechanism when unit A does not wait for unit B.  "Ordering" is a separate thing from "depending".  But then, `Requisite` may need to be two different things.

> try to describe what you are trying to achieve in general terms (without referring
> to any specific systemd directives) and do it on mailing list.

 I think that you have illuminated that a `Requisite` which waits for a `DefaultTimeoutStartSec` time in the future is a different thing from a `Requisite` that applies immediately and refers to the current time.  And then, I have an idea.

The simplest approach could be to only address the behavior of `BindsTo=B.device`, and then sort of ignore the other issues.

The same time distinction would need to apply to `BindsTo`, the "selection time" versus the "post-ExecStart time".  And, we might need to call this two different things.  In particular, when unit A is self-referential and instantiates a B.device, the `ExecStart` commands would need to be run before unit A checks for B.device.

Hmm, hmm - maybe there only needs to be a new "`Prerequisite`" dependency, which always only applies to the "selection time"/"current time"?  And then, we could document that `BindsTo` start and `Requisite` always only apply to a "post-ExecStart time"/"time in the future"?  That might also be the least disruptive and most backward-compatible approach to the nonexistent device issue.

In my use case, the unit file which creates the bridge and bond devices must also "clean-up-from" the destruction of these devices and uses the `BindsTo` dependency.  But the `ExecStart` commands must run before checking that the device exists.  In this case there would be NO `Prerequisite=B.device`.  On the other hand, the unit file which enslaves the physical devices to the bridge and bond devices must not be activated until these physical devices exist first.  In that case there could be a `Prerequisite=B.device` dependency added, to block activation of the `ExecStart` commands until B.device appears.  There would be no need for the `Requisite` dependency here.

Perhaps `Prerequisite=B.device` need only change the initial state of the `BindsTo` mechansim, which already is able to wait for the appearance of a B.device.

`Prerequisite` and `Requisite` would illuminate the distinction between the pre`ExecStart` time and the post`ExecStart` time.  And, with that distinction, would a use case exist for the operation of a `Requisite` dependency that does *not* imply the effect of `After=`?

If not, then `Requisite` might also automatically set `After=`, just because that's the only thing that makes sense, and there would be no reason to write an explicit `After=` into the unit file to make `Requisite` function correctly.

**Comment by arvidjaar** [comment]
> With `After=B.device` in unit file A, then unit A waits for B.device.  Would you agree?

No.

**Comment by thx1111** [comment]
> In your case something else already waits for device, which causes JOB_VERIFY_ACTIVE
> to be subsumed by this waiting job.

I have the enslaving unit file `WantedBy` the bridge and bond master unit file, as well as `BindsTo` the physical device, so that the individual physical slaves can be manually released from the master, but then restarting the master unit file causes all the slave devices to be enslaved again.

The master unit file cannot know in advance which physical devices are or are not currently plugged, but the resulting `Wants` on a nonexistent device will still invoke the enslaving unit file for that physical, but possibly nonexistent, device.

The master unit file is ordered before a target unit file, and the enslaving unit file is ordered after this target unit file, so that no attempt is made to enslave a physical device to a nonexistent master device.  With the enslaving unit file ordered after B.device, the enslaving unit is ordered after all of the master unit file, this target unit file, and B.device.
The master unit file and the enslaving unit file are also both ordered before network.target.

So network.target is waiting on the enslaving unit, which is waiting on B.device.

If I understand, JOB_VERIFY_ACTIVE could be subsumed into the network.target job, which is then effectively waiting on the nonexistent device.

Is that more correct?

This produces the temporary "A start job is running for B.device" journal message, but systemd does not disclose *which* start job is running for B.device.
